File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/UseItemIcon.vb
Imports Sakuno.ING.Game.Models.Knowledge

Namespace Controls
    Public Class UseItemIcon
        Inherits UserControl

        Private ReadOnly _image As Image
        Public Sub New()
            _image = New Image()
            Content = _image
        End Sub

        Public Shared ReadOnly IdProperty As DependencyProperty =
            DependencyProperty.Register(NameOf(Id), GetType(KnownUseItem), GetType(UseItemIcon),
                                        New PropertyMetadata(New KnownUseItem, Sub(d, e) CType(d, UseItemIcon).OnIdChanged(CType(e.NewValue, KnownUseItem))))

        Public Property Id As KnownUseItem
            Get
                Return CType(GetValue(IdProperty), KnownUseItem)
            End Get
            Set(value As KnownUseItem)
                SetValue(IdProperty, value)
            End Set
        End Property

        Private Sub OnIdChanged(value As KnownUseItem)
            Dim intId As Integer = value
            If DesignMode.DesignModeEnabled Then
                _image.Source = New BitmapImage(New Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/UseItem/{intId}.png"))
            Else
                Dim stringId = "UseItemIcon_Bitmap_" & intId
                Dim source As Object = Nothing
                If Application.Current.Resources.TryGetValue(stringId, source) Then
                    _image.Source = DirectCast(source, BitmapImage)
                Else
                    Dim s = New BitmapImage(New Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/UseItem/{intId}.png"))
                    Application.Current.Resources.Add(stringId, s)
                    _image.Source = s
                End If
            End If
        End Sub
    End Class
End Namespace

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/UseItemIcon.cs
using System;
using Sakuno.ING.Game.Models.Knowledge;
using Windows.ApplicationModel;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media.Imaging;

namespace Sakuno.ING.Views.UWP.Controls
{
    public sealed class UseItemIcon : UserControl
    {
        private readonly Image _image = new Image();
        public UseItemIcon() => Content = _image;

        public static readonly DependencyProperty IdProperty
            = DependencyProperty.Register(nameof(Id), typeof(KnownUseItem), typeof(UseItemIcon),
                new PropertyMetadata((KnownUseItem)0, (d, e) => ((UseItemIcon)d).Update()));
        public KnownUseItem Id
        {
            get => (KnownUseItem)GetValue(IdProperty);
            set => SetValue(IdProperty, value);
        }

        private void Update()
        {
            int intId = (int)Id;
            if (DesignMode.DesignModeEnabled)
                _image.Source = new BitmapImage(new Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/UseItem/{intId}.png"));
            else
            {
                string stringId = "UseItemIcon_Bitmap_" + intId;
                if (Application.Current.Resources.TryGetValue(stringId, out object source))
                    _image.Source = source as BitmapImage;
                else
                {
                    var s = new BitmapImage(new Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/UseItem/{intId}.png"));
                    Application.Current.Resources[stringId] = s;
                    _image.Source = s;
                }
            }
        }
    }
}
